docs/development/sync-protocol.md

```diff
@@ -0,0 +1,431 @@
+---
+title: Santa Sync Server Protocol
+parent: Development
+mermaid: true
+---
+
+{% if page.mermaid %}
+  {% include mermaid.html %}
+{% endif %}
+
+# Overview
+
+This documents describes the protocol between Santa and the Sync Server aka the Sync Protocol. Implementors should be able to use this to create their own Sync Servers.
+
+## Background
+
+Santa can be run and configured with a sync server. This allows an admin to
+easily configure and sync rules across a fleet of macOS systems.  In addition to
+distributing rules using a sync server enables an admin to override some local
+configuration options e.g. lockdown mode on both a fleetwide and per host basis.
+
+# The Santa Sync Server Protocol 50,000ft View
+
+The Sync Protocol is an HTTP/JSON Restful protocol. As such it is
+assumed that both the server and client add `Content-Type` headers are set to
+`application/json`. 
+
+The Sync Protocol is client initiated and consists of 4 request-response
+transactions called stages, `preflight`, `eventupload`, `ruledownload` and `postflight`. 
+A sync may consist of all 4 stages, or just the `ruledownload` stage.
+
+| Stage | What it Does |
+|---|---|
+| **Preflight** | Report current Santa settings and machine attributes to Sync Server & retrieve configuration settings |
+| **Event Upload** | Report new blockable events to the Sync Server |
+| **Rule Download** | Retrieves new rules |
+| **Postflight** | Reports stats | 
+
+If the server returns a value other than 200 for any stage than the sync stops and the next stage is not performed.
+
+At a high level this looks like the following sequence:
+
+```mermaid
+sequenceDiagram
+   client ->> server: POST /preflight/<machine_id>
+   server -->> client: preflight response (200)
+   client ->> server: POST /eventupload/<machine_id>
+   server -->> client: eventupload response (200)
+   client ->> server: GET /ruledownload/<machine_id>
+   server --> client: ruledownload response (200)
+   client ->> server: POST /postflight/<machine_id> request
+   server -->> client: postflight response (200)
+```
+
+Where `<machine_id>` is a unique string identifier for the client by default
+Santa makes this is the string `IOPlatformUUID` from IOKit's
+`IOPlatformExpertDevice` aka the hardware UUID.
```

Comment:
I think "Santa makes this the hardware UUID" is enough.

Reply:
Though I think you should mention that this is configurable using the MachineID* keys and link to the configuration doc